### DISCUSSION TITLE
Connection lists, Favourites, Bookmarks

### DISCUSSION THREAD

**Discussion by geotechfirst** [jamulussoftware/jamulus#1774] [Ideas]
Following @ann0see's suggestion this is thread to come to a common understanding about a requirement that might benefit the wider Jamulus community.
From my perspective I needed the ability to connect to a set of servers in quick succession with minimal hassle.
Background: 
During WorldJam performances performers need to change servers multiple times.
As a performer I found this to be not convenient. As you are already quite nervous about your performance.
I understand that this feature might not be useful to the wider community so took on board comments from others and made a prototype here:
https://geotechfirst.github.io/bookmarks/
and submitted a PR https://github.com/jamulussoftware/jamulus/pull/1728

In summary my bare bone requirement is:
- ability load a list of servers from a file
- ability to connect to these servers quickly and conveniently
 
I am happy to go back to the drawing board to work on a common feature that may become useful for others.
@dcorson-ticino-com Also has a PR https://github.com/jamulussoftware/jamulus/pull/1552  I let him tell what the driver behind that.

**Comment by dcorson-ticino-com** [comment]
Hi @geotechfirst,
My Favorite/Bookmarks was thought to be more general.  I see many people who are overwhelmed by the quantity of servers available and who will be only using a small number for rehearsals.  In such a case a simple list ordered with the last used at the top is a good solution.  Making the last used also selected when the window is opened connecting to the last used is simply open the window and click connect.  That compared with open the window, wait a bit for the list to stabilize, find the server you want to use in the list, select it and click connect.

For your use my solution will only reduce the number of servers in the list, but not put them in a certain order.  To go in this direction I added the possibility to sort by last used or Directory, thinking you could use the  directory sort as a basis and change that to the WJ functionality you need.  You would not need to worry about all the behind the scenes management, de-referencing servers for example, and could add your code changing the Directory sort (which for itself is not all that useful) to WJ functionality.  I would see management like deleting and reordering on the RMB and why not add a button for "connect to next" in the window. I save the Favorites in the Jamulus ini file, but that could also be a separate file.  Note that servers are added by noting the presently connected server so all the needed info is taken internally, no manual lookup of anything is needed.

What I would not do is remove the simple last used Favorites list as I think that has a much bigger user base than WJ. And I have no particular love of the term Favorites, Bookmarks sounds just as good to me too.

As I do not know the details of the needs for WJ I will let you take over and run with this.  Let me know if you need any clarification.

**Comment by passing** [comment]
Hi @geotechfirst,

If I understand the problem that you are trying to solve right, you may be able to solve it with a completely different approach:
What I get is that you don't want all WJ participants to see all WJ servers, but only the servers relevant for each participant.
You could run virtual directory servers, that return the server lists as needed. Then each participant just needs to configure a different directory server and will get the relevant servers. That would even have a few advantages over the proposed solution:
- You would see in the server list of Jamulus who is connected to each server before you enter
- It is possible to change the list of servers without requiring each participant to download it again

With the Jamulus python library, that should be fairly easy:
https://github.com/passing/jamulus-python/blob/master/central_server.py#L56-L71
You could either implement multiple python based directory servers, each returning a different server list. Or you make one server with multiple endpoints (listening to different ports), but that would make the implementation more complex.

**Reply by hsweet** [reply]
Would a custom Central/directory server work? It's already built into Jamulus. If all the WJ servers were pointed to custom everyone participating can see only the other related servers and who is in them.  I just set up that scenario for a small festival and have a few scripts here that make that process automatic. https://github.com/hsweet/jamulus.   Outside of that I like the bookmark idea too.

**Reply by dcorson-ticino-com** [reply]
There is already a WorldJam custom Directory server (worldjam.vip), but as it points to all the servers around the world for one particular musician it shows too many.  @geotechfirst 's idea is to extract only the servers that one needs and place them in the order that they will be used.  For example: "Warm-up Room EU" -> "Waiting Rm 1" -> "Studio London", then with a button go from one to the next when called for.
This is quit possibly a good thing for the WorldJam concerts and the attendant nervousness before a concert, but very WorldJam specific.
